test/LRUCacheTest.cpp

```diff
@@ -42,6 +62,64 @@ TEST(LRUCacheTest, testTryEmplace) {
   *emplaced += "bar";
   ASSERT_EQ(*cache["5"], "foobar");
 }
+// _____________________________________________________________________________
+TEST(LRUCacheTest, testTryEmplaceWithDrop) {
+  LRUCache<string, string> cache(3);
+  cache.insert("1", "x");
+  cache.insert("2", "xx");
+  ASSERT_FALSE(cache.tryEmplace("2", "foo").first);
+  ASSERT_EQ(*cache.tryEmplace("3", "bar").first, "bar");
+  auto emplaced = cache.tryEmplace("4", "foo").first;
+  ASSERT_TRUE(emplaced);
+  *emplaced += "bar";
+  ASSERT_EQ(*cache["4"], "foobar");
+  ASSERT_FALSE(cache["1"]);      // Dropped oldest
+  ASSERT_EQ(*cache["2"], "xx");  // Kept second oldest
+}
+// _____________________________________________________________________________
+TEST(LRUCacheTest, testTryEmplacePinnedSimple) {
+  LRUCache<string, string> cache(3);
+  ASSERT_EQ(*cache.tryEmplacePinned("pinned", "bar").first, "bar");
+  cache.insert("1", "x");
+  cache.insert("2", "xx");
+  cache.insert("3", "xxx");
+  cache.insert("4", "xxxx");
+  ASSERT_EQ(*cache["pinned"], "bar");  // pinned still there
+  ASSERT_FALSE(cache["1"]);            // oldest already gone
+  ASSERT_EQ(*cache["2"], "xx");  // not dropped because pinned not in capacity
+  ASSERT_EQ(*cache["pinned"], "bar");  // pinned still there
+}
+// _____________________________________________________________________________
+TEST(LRUCacheTest, testTryEmplacePinnedExisting) {
+  LRUCache<string, string> cache(2);
+  cache.insert("1", "x");
+  cache.insert("2", "xx");
+  // tryEmplacePinned on an existing element doesn't emplace but it does pin
+  ASSERT_FALSE(cache.tryEmplacePinned("2", "yy").first);
```

Comment:
This test should check that the newly pinned element is correctly removed from the normal cache? Because I think lines 126 and 127 aren't currently tested:
https://github.com/ad-freiburg/QLever/blob/fd51501706170d615a230da9bf7fa59c74c48ccc/src/util/LRUCache.h#L123-L128
Because of the insertions of 4 and 5 which replace the whole contents of the normal cache before accessing anything there.

Reply:
Hmm, well the pinned cache sits transparently in front of the normal cache so any test if it's not in there anymore would need internal access. Am I missing something here?

Reply:
That's right, but maybe an internal test should be added to check that an element does not belong to both parts of the cache. The code is currently there but I can imagine that these lines could be missed during some work there and then the cache is one spot smaller (until the value drops out naturally) as the pinned result is always accessed first.

Reply:
I hope you don't mind that I tried to add the test in 6533484 to this PR.

Reply:
Looks great